spec/facebook/messenger/incoming/common_spec.rb

```diff
@@ -64,72 +64,117 @@ class Dummy
 
   describe '.typing_on' do
     let(:access_token) { 'access_token' }
+    let(:app_secret_proof) { 'app_secret_proof' }
 
     it 'sends a typing indicator to the sender' do
       expect(Facebook::Messenger.config.provider).to receive(:access_token_for)
         .with(subject.recipient)
         .and_return(access_token)
+      expect(Facebook::Messenger.config.provider).to(
+        receive(:app_secret_proof_for)
+          .with(subject.recipient)
+          .and_return(app_secret_proof)
+      )
 
       expect(Facebook::Messenger::Bot).to receive(:deliver)
         .with({
                 recipient: subject.sender,
                 sender_action: 'typing_on'
-              }, access_token: access_token)
+              }, access_token: access_token, app_secret_proof: app_secret_proof)
+
+      subject.typing_on
+    end
+  end
+
+  describe '.typing_on with app_secret_proof disabled' do
```

Comment:
I see that it's single method you test when proof is disabled. So for typing_on both cases are covered.
All other methods are tested only when proof is enabled.

Is there any reason behind it?

Reply:
sure, I thought it provided adequate test coverage as we're really testing the base class so all the other code paths are the same, I didn't really want to confuse the tests with a lot of noise.

Reply:
Just sharing my view on it, not forcing ;)

The only one difference is the data needs to be delivered, so by extracting delivery to shared examples group we could cover both cases for all methods, not for only single one.

Advantages that duplicates are removed + you explicitly say that all methods work similar.

So this is the set of tests covering 'typing_on', 'typing_off', 'mark_seen', 'reply' methods (i didn't run it, sharing the idea).

 ```suggestion
  describe 'replying to message' do
    let(:access_token) { 'access_token' }
    let(:app_secret_proof) { 'app_secret_proof' }

    before do
      allow(Facebook::Messenger.config.provider).to receive(:access_token_for)
        .with(subject.recipient).and_return(access_token)

      allow(Facebook::Messenger.config.provider).to receive(:app_secret_proof_for)
        .with(subject.recipient).and_return(app_secret_proof)
    end

    shared_examples_for 'payload delivery' do |delivery_method|
      shared_examples_for 'request execution' do
        it 'delivers the payload' do
          expect(Facebook::Messenger::Bot).to receive(:deliver)
            .with(payload, access_token: access_token, app_secret_proof: app_secret_proof)

          subject.public_send(delivery_method)
        end
      end
      
      context 'when app secret proof enabled' do
        include_examples 'request execution'
      end

      context 'when app secret proof disabled' do
        let(:app_secret_proof) { nil }

        include_examples 'request execution'
      end
    end

    it_behaves_like 'payload delivery', :typing_on do
      let(:payload) do
        {
          recipient: subject.sender,
          sender_action: 'typing_on'
        }
      end
    end

    it_behaves_like 'payload delivery', :typing_off do
      let(:payload) do
        {
          recipient: subject.sender,
          sender_action: 'typing_off'
        }
      end
    end

    it_behaves_like 'payload delivery', :mark_seen do
      let(:payload) do
        {
          recipient: subject.sender,
          sender_action: 'mark_seen'
        }
      end
    end

    it_behaves_like 'payload_delivery', :reply do
      let(:payload) do
        {
          recipient: subject.sender,
          message: { text: 'Hello, human' },
          message_type: Facebook::Messenger::Bot::MessageType::RESPONSE
        }
      end
    end
  end

```

Reply:
Seems DRY-er. Changed in a force push to the last commit.